pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -305,13 +312,55 @@ func (m *KV) GetCodec(codecID string) codec.Codec {
 
 // GetListeningPort returns port used for listening for memberlist communication. Useful when BindPort is set to 0.
 func (m *KV) GetListeningPort() int {
-	return int(m.memberlist.LocalNode().Port)
+	return int(m.getMemberlist().LocalNode().Port)
 }
 
-// JoinMembers joins the cluster with given members.
+// JoinMembers joins the cluster with given members. If the first attempt fails (e.g. no members reached) and
+// immediate abort is not set JoinMembers will retry with backoff up to max retries configured.
 // See https://godoc.org/github.com/hashicorp/memberlist#Memberlist.Join
 func (m *KV) JoinMembers(members []string) (int, error) {
-	return m.memberlist.Join(members)
+	reached, err := m.getMemberlist().Join(members)
+	if err != nil && m.cfg.AbortIfJoinFails {
+		_ = m.getMemberlist().Shutdown()
+		return 0, fmt.Errorf("abort joining memberlist cluster: %w", err)
+	}
+
+	if reached > 0 {
+		level.Debug(util.Logger).Log("msg", "joined memberlist cluster", "node", m.cfg.NodeName, "reached", reached)
+		return reached, nil
+	}
+
+	if m.cfg.MaxJoinRetries > 0 {
+		go m.retryJoinMembersWithBackoff(members)
+	}
+
+	return 0, nil
```

Comment:
I suggest we move this logic to `NewKV` for now. When/if https://github.com/cortexproject/cortex/pull/2713 is merged, we can move it to `starting method. WDYT?